bauhaus/constraint_builder.py

```diff
@@ -128,40 +138,63 @@ def build(self, propositions) -> 'NNF':
             A built NNF constraint
 
         """
+        if self._groupby:
+            assert len(propositions) == 1, "Should only be using groupby on a single class"
```

Comment:
len(propositions) == number of classes decorated with @proposition for an encoding. 

I think we may need a different condition, or else this will only work for an Encoding with a single @proposition

Reply:
Good catch, and a point I was confused about. I'd assumed a `_ConstraintBuilder` object was built once for each decoration that takes place. How many of those objects are actually created during the encoding?

Reply:
You're right, just one _ConstraintBuilder for each constraint decorator or function call!

I think if groupby must only be applied to a single class, we don't need a check for it since it must already be true if the ConstraintBuilder object has a self._func and if not ismethod(self._func) (ismethod is from utils.py)

Reply:
I think other than that condition, everything else looks great!

Reply:
Hrmz...still not sure I'm fully understanding the scenario. The check was there in case someone attempted a groupby= kwarg on a method annotation or direct constraint addition. Would either of those cases play a role here?

Reply:
**\*nudge\***

Reply:
hey! happy holidays :) 

right, I think I need to clarify too! I was thinking the following:

- **groupby with class decorator**: only applicable over a single class (the defined self._func). then it follows we are only obtaining the propositional variables stored for that single class.
- **groupby with a method**: I don't fully recall the use case for this, but we can retrieve the propositional variables from the class containing the decorated method, so there are no concerns here for applying groupby over multiple classes of propositional variables.
- **direct constraint addition**: I think here we need to be checking that a single argument is passed and that it is a class or method function. That way, we know we are only applying groupby over a single class' propositional variables.

is this in line with what you had for the check/scenario?

Reply:
Actually, it was just the top. I hadn't considered the interpretation of what using `groupby` would mean on a method. The third option I wanted to avoid as well, but assumed that's a different call path (i.e., we don't even hit this code if it's a direct constraint).

If it's on a method, then that will be called for each instance of the class -- unclear would `groupby` should do in that case.

Oh, and happy new years!!

Reply:
so to prevent someone from using groupby on a method (since it's a bit redundant), we can check the following during constraint creation in core.py (i.e. [this line](https://github.com/QuMuLab/bauhaus/blob/master/bauhaus/core.py#L322) ) :
```
from utils import ismethod
if self._groupby and ismethod(self.func):
    # raise exception: ask user to decorate class instead of method
```
since this would be before constraint building, we actually wouldn't need any checks in the build function. I'll look into this a bit more though with some tests to make sure it checks out for all cases.

Reply:
Hope I've handled this in the latest commit!